src/client.ts

```diff
@@ -1,46 +1,65 @@
-/// `client.ts` is the main entry point into the client code
-///
-/// run with `npx ts-node src/client.ts` to start up our client
-
-import { io, Socket } from "socket.io-client"
-import { ClientEvents, ServerEvents } from "./server/events"
+import WebSocket from "ws";
 import config from "./config"
 
-const socket: Socket<ServerEvents, ClientEvents> = io("http://localhost:3000")
-
-socket.on("connect", () => {
-    console.log(`connect ${socket.id}`)
-
-    let ryder_port: string
-
-    if (!config.ryder_port) {
-        const args = process.argv.slice(2)
-        const ryder_port_idx = args.findIndex(arg => arg === "--ryder_port")
-        if (ryder_port_idx !== -1 && ryder_port_idx !== args.length) {
-            ryder_port = args[ryder_port_idx + 1]
-        } else {
-            console.log({ args })
-            console.log("Missing RYDER_PORT environment variable.")
-            process.exit(0)
-        }
-    } else {
-        ryder_port = config.ryder_port
-    }
-
-    // this is opening ryder-serial at `ryder_port`
-    socket.emit("serial:open", { port: ryder_port, options: { debug: true } }, res => {
-        console.log(res)
-    })
-})
-
-socket.on("serial:opened", () => {
-    console.log("serial:opened")
-})
-
-socket.on("connect_error", err => {
-    console.log(`connect_error due to ${err.message}`)
-})
-
-socket.on("disconnect", reason => {
-    console.log(`disconnect due to ${reason}`)
-})
+type Options = { [key: string]: any };
+
+export class RyderWebsocketClient {
+	protected ws!: WebSocket;
+	protected options: Options;
+	protected connecting = false;
+
+	protected call_id = 0;
+	protected queue: Array<[Function, Function]> = [];
```

Comment:
Not sure if it's necessary because it'll always be the same two functions pushed onto the queue. I'll leave it up to you, as I don't know if there's a more formal TS way to represent resolve and reject.

Reply:
For `Resolve` and `Reject` types: 

```typescript
type Resolve = (value: void | PromiseLike<void>) => void;
type Reject = (reason?: any) => void;
```

But I'm fine with leaving `queue` as a simple array, actually (not its own implementation)